### PR TITLE
PAL Module/shutdown cleanup.

### PR DESCRIPTION
**Pull request by mikem8361**
Added a PAL_Shutdown function that cleans/shutdowns the PAL without exiting/terminating
the process doing exactly what ExitProcess/TerminateProcess before exiting the process.

Removed the PAL_InitializeCoreCLR export. The new coreclr_* hosting apis should be used.

Moved the "is module manager initialized" check (exe_module.prev != NULL) inside the module
lock to prevent races during shutdown in LOADCallDllMain.

Removed LOADFreeModules so we never try to call the DLL_PROCESS_DETACH handlers. Most of this
cleanup happened because I was trying to getting the DLL_PROCESS_DETACH notifications working,
but there were too many opportunities for deadlock. Emulating Windows behavior isn't that
important and in this case will cause more problems that it fixes.

Removed PAL_RegisterLibrary*/PAL_UnregisterLibrary. They were only used in one place to load modules
that don't exist anymore (rotor_palrt and sscoree).

Renamed LOAD_SEH_CallDllMain to LoadCallDllMainSafe (we don't usually use _ in names) and use this
function in all the places we call the DllMain handler. This combines some code in a common place
and protects all DllMain calls from exceptions.

Source code formatting cleanup and moved internal functions around to be next to the rest of the
internal functions.

Initialize pDllMain to NULL in LOADAllocModule.

Get the DllMain function address into LOADAddModule so that it is initialized in the PAL_RegisterModule path.

Changed PAL_UnregisterModule not to call the DllMain function.

### CODE REVIEW

**Comment by mikem8361** [comment]
@janvorli, @jkotas and optionally @sergiy-k could you review?

File: src/dlls/mscoree/unixinterface.cpp [hunk 1]
```diff
@@ -178,13 +178,13 @@ static void ExtractStartupFlagsAndConvertToUnicode(
 //
 extern "C"
 int coreclr_initialize(
-            const char* exePath,
```
Comment (janvorli) [inline src/dlls/mscoree/unixinterface.cpp]: These indents were intentional to not to align the parameters with the name of the function / function code lines. Could you please revert this change?

File: src/pal/src/init/pal.cpp [hunk 2]
```diff
@@ -742,32 +741,17 @@ PAL_EntryPoint(
 
 /*++
 Function:
-  PALCommonCleanup
-
-Utility function to prepare for shutdown.
+  PAL_Shutdown
 
+Abstract:
+  This function is the called. It shuts down the PAL WITHOUT exiting the current process.
```
Comment (janvorli) [inline src/pal/src/init/pal.cpp]: I don't understand the first sentence of the comment.

File: src/pal/src/init/pal.cpp [hunk 3]
```diff
@@ -742,32 +741,17 @@ PAL_EntryPoint(
...
 --*/
-void 
-PALCommonCleanup()
+void
+PALAPI
+PAL_Shutdown(
```
Comment (janvorli) [inline src/pal/src/init/pal.cpp]: Why do we need the PAL_Shutdown?

File: src/dlls/mscoree/unixinterface.cpp (hunk 1 above)
Comment (mikem8361) [inline src/dlls/mscoree/unixinterface.cpp]: I thought the coding convention was 8 space indentation.  At least a majority of the code uses 8 spaces when the function parameters are put on a separate line.

But I'll revert these changes if you want.

mikem

File: src/pal/src/init/pal.cpp (hunk 3 above)
Comment (mikem8361) [inline src/pal/src/init/pal.cpp]: I put it in at first so I could send the DLL_PROCESS_DETACH notifications on shutdown but that has too many problems.  I left it in for symmetry with PAL_Initialize and for any future cleanup, etch.  It currently cleans up exactly what ExitProcess/TerminateProcess does.

mikem

File: src/dlls/mscoree/unixinterface.cpp (hunk 1 above)
Comment (mikem8361) [inline src/dlls/mscoree/unixinterface.cpp]: I meant "4" spaces below not "8".

Comment (janvorli) [inline src/dlls/mscoree/unixinterface.cpp]: Regarding the parameters, it really depends on where you look in the codebase. Some code uses what you said, some aligns all parameters with the opening brace, some uses what I've used.
It would be nice if you could revert it here.
